pints/residuals_diagnostics.py

```diff
@@ -11,6 +11,311 @@
+    n_bins
+        Optional int value (greater than zero) giving the number of bins into
+        which to divide the time series. By default, it is fixed to 25.
+    calculate : function
+        What value to calculate within each bin. This function should take as
+        input a numpy array of residuals within a bin and return a scalar value
+        which will be added to the plot.
+    label : str
+        A label to put on the y axis of the plot, describing what function of
+        the binned residuals is being plotted.
+    ylim : tuple
+        Optional bounds for the y-axis of the plot.
+    draw_horizontal : bool
+        Whether or not to draw a horizontal line at y=0 on the plot. This line
+        may be desired when zero is an important reference point for the value
+        being calculated in each bin.
+    """
+    import matplotlib.pyplot as plt
+
+    times = problem.times()
+
+    # Get the number of problem outputs
+    n_outputs = problem.n_outputs()
+
+    # Get the matrix of residuals values
+    residuals = calculate_residuals(parameters, problem, thinning=thinning)
+
+    # Get the posterior median residuals
+    residuals = np.median(residuals, axis=0)
+
+    n_bins = int(n_bins)
+    if n_bins < 1:
+        raise ValueError('n_bins must be an integer value greater than zero.')
+    if n_bins > len(times):
+        raise ValueError('n_bins must not exceed the number of time points')
+
+    # Make the figure, with one axes for each output
+    fig = plt.figure()
+    for output in range(n_outputs):
+        ax = fig.add_subplot(n_outputs, 1, output + 1)
+
+        # Get the residuals along this output
+        e = residuals[output, :]
+
+        # Divide the residuals into bins
+        binned_data = np.array_split(e, n_bins)
+        binned_times = np.array_split(times, n_bins)
+
+        # Calculate lag 1 autocorrelation and time in each bin
+        bin_values = []
+        bin_times = []
+        for data, t in zip(binned_data, binned_times):
+            bin_values.append(calculate(data))
+            bin_times.append(np.mean(t))
+
+        # Plot the binned data
+        ax.plot(bin_times, bin_values, 'o-', color='red')
+
+        if ylim is not None:
+            ax.set_ylim(*ylim)
+
+        ax.set_xlabel('Time')
+        ax.set_ylabel(label)
+
+        ax.set_title('Output %d' % (output + 1))
+
+        if draw_horizontal:
+            # Draw a horizontal line at 0 autocorrelation
+            ax.axhline(0, color='C0', zorder=-10)
+
+    return fig
+
+
+def plot_residuals_distance(parameters, problem, thinning=None):
+    r"""
+    Plot a distance matrix of the residuals.
+
+    Given a time series with observed residuals
+
+    .. math::
+        e_i = y_i - f(t_i; \theta)
+
+    this function generates and plots the distance matrix :math:`D` whose
+    entries are defined by
+
+    .. math::
+        D_{i, j} = |e_i - e_j|
+
+    The plot of this matrix may be helpful for identifying a time series with
```

Comment:
Isn't it more that banded noise structure suggests that autocorrelation is temporally varying?

Reply:
I think I can improve the wording here. But wont the matrix still appear banded even for a level of autocorrelation which is constant over time? I'm thinking of one example I did in the [notebook](https://github.com/pints-team/pints/blob/b6e8c60d6e1d80a71cda55acef2b734f28768e4f/examples/plotting/residuals-autocorrelation-diagnostics.ipynb) (cell 6), where the distance matrix still appears distinctively banded for AR(1) noise with constant rho=0.85.